ksql-api/src/test/java/io/confluent/ksql/api/perf/PullQueryRunner.java

```diff
@@ -0,0 +1,153 @@
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.perf;
+
+import static io.confluent.ksql.api.perf.RunnerUtils.DEFAULT_COLUMN_NAMES;
+import static io.confluent.ksql.api.perf.RunnerUtils.DEFAULT_COLUMN_TYPES;
+import static io.confluent.ksql.api.perf.RunnerUtils.DEFAULT_ROW;
+
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.api.impl.VertxCompletableFuture;
+import io.confluent.ksql.api.server.BufferedPublisher;
+import io.confluent.ksql.api.server.InsertResult;
+import io.confluent.ksql.api.server.InsertsStreamSubscriber;
+import io.confluent.ksql.api.spi.Endpoints;
+import io.confluent.ksql.api.spi.QueryPublisher;
+import io.vertx.core.Context;
+import io.vertx.core.WorkerExecutor;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.client.HttpResponse;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Semaphore;
+import org.reactivestreams.Subscriber;
+
+public class PullQueryRunner extends BasePerfRunner {
+
+  public static void main(String[] args) {
+    new PullQueryRunner().go();
+  }
+
+  private static final String DEFAULT_PULL_QUERY = "select * from foo where rowkey=123;";
+  private static final JsonObject DEFAULT_PULL_QUERY_REQUEST_BODY = new JsonObject()
+      .put("sql", DEFAULT_PULL_QUERY)
+      .put("properties", new JsonObject());
+  private static final List<GenericRow> DEFAULT_ROWS = generateResults();
+  private static final int MAX_CONCURRENT_REQUESTS = 100;
+
+  private PullQueryEndpoints pullQueryEndpoints;
+
+  @Override
+  protected void configure() {
+    this.pullQueryEndpoints = new PullQueryEndpoints();
+    setNumWarmupRuns(5).setNumRuns(5).setRunMs(10000).setEndpoints(pullQueryEndpoints);
+  }
+
+  @Override
+  protected void run(long runMs) throws Exception {
+    Semaphore sem = new Semaphore(MAX_CONCURRENT_REQUESTS);
+
+    long start = System.currentTimeMillis();
+
+    do {
+
+      sem.acquire();
+
+      VertxCompletableFuture<HttpResponse<Buffer>> vcf = new VertxCompletableFuture<>();
+
+      client.post(8089, "localhost", "/query-stream")
+          .sendJsonObject(DEFAULT_PULL_QUERY_REQUEST_BODY, vcf);
+
+      vcf.thenAccept(resp -> {
+        count();
+        sem.release();
+      });
+
+    } while (System.currentTimeMillis() - start < runMs);
+  }
+
+  @Override
+  protected void endRun() throws Exception {
+    pullQueryEndpoints.closePublishers();
+
+    Thread.sleep(500);
```

Comment:
As above, why the sleep with a magic number of ms?  Such sleeps in tests are a code smell and often result in intermittent failures.